tp26/README.md

```diff
@@ -0,0 +1,140 @@
+1. The exposed DWN, named in the `grantedFor` property. This is the DWN to which the grantee is given access.
+2. The grantor, named in the `grantedBy` property. The grantor must retain the grant in order to create a `PermissionsRevoke` in the future.
+3. The grantee, named in the `grantedTo` property. The grantee includes the grant in messages to the exposed DWN.
+In most scenarios, the DWN belongs to the grantor. However, for delegated grants, the DWN owner and the grantor are separate entities, so separate fields are necessary to differentiate them.
+
+Why is `grantedFor` needed? Consider the alternative. If `grantedFor` was not present, then it would be strange for the grantor and grantee to store the `PermissionsGrant` in their own DWN. The grantor's and grantee's DWNs would have no way to discern that "this grant is meant to give grantee access to someone else's DWN, not to give someone else access to mine." One option is for the grantor and grantee to store the grant in their respective web5 wallets. Storing in their DWN is better though, because their DWNs have built-in message validation and message validation.
+
+
+## Scope and Conditions
+The descriptors of `PermissionsRequest` and `PermissionsGrant` both contain the `scope` and `conditions` properties. `scope` is required, and `conditions` is optional. Some `scope` and `conditions` options may be relevant only for a few DWN message types.
+
+The `scope` property sets the boundaries on what abilities the grantee has. `scope` specifies a single DWN interface and method as well as several optional fields. The abilities granted by the `PermissionsGrant` are the intersection of properties within `scope`. In other words, whenever a grant is used, the action taken must follow all of the rules outlined in the `scope`.
+
+The `conditions` property describes certain features about messages which use the grant. For example, if a grant has the `encryption` condition set to `required` then all records written using that grant must be encrypted.
+
+### Delegation
+The `delegation` condition contains enough complexity that it deserves a separate technical proposal. There are mentions of delegates and delegation in this TP, but its nuance is not fully captured. For the curious, delegation allows the following scenario: if Alice gives Bob a grant with `delegation` enabled, then Bob can issue grants of the same or narrower scope to Carol.
+
+### Type Definitions
+See the type definitions below for a list of `scope` and `conditions` options. More options may be added in the future.
+```typescript
+type PermissionsScope = {
+  interface: DwnInterface;
+  method: DwnMethod;
+  // The grantee may access records with matching `schema`
+  schema?: string;
+  // The grantee may access protocols or records with matching protocol URI.
+  protocol?: string;
+  // The grantee may access a specific records with matching recordId.
+  recordIds?: string[];
```

Comment:
Similar questions to snapshots:

1. Are these mutually exclusive?
1. Are these "enough" for real-world use cases? For instance: can you grant access to a certain protocol context?
1. What does it mean to grant write access to a protocol? It's strange because it implies protocol violation is imminent and we are okay with that...?!

Reply:
1. No, these are not mutually exclusive. Including `schema` and `protocol` for a grant allowing `RecordsWrites` would allow someone to write records to a protocol with the specified schema. In practice, it may not be useful to have both of those, but they are not technically exclusive.
2. I just added `contextId`, which covers enough real world use cases for now.
3. Amended based on conversations with the team this past week. A `PermissionsGrant` can allow an entity access to protocol records that the protocol definition would not otherwise allow them access to. However, for `RecordsWrites` the record must conform to the type laid out in the protocol definition. In other words, an entity can gain access to protocol records EITHER by following the rules of the protocol definition `structure` OR by presenting a `PermissionsGrant`.

Reply:
Amen to 3.

On 1:

While this is not a must for permissions feature, the "mutually exclusiveness" is there for the snapshot feature, and in those discussions we established that having "compatible" if not the same scoping structure between the two features (as well as protocols feature) is quite desirable. An example undesirable scenario: Alice is granted a permission to write/update records with schema X which also INCLUDES records under protocols, revoking her permission would require potentially many snapshots to fully restrict her access.

On 2:
I'm fine with `contextId` being the only additional property added for now. But want to call out that one might also want to grant permission to objects under a certain protocol path. This kinda reiterates the usefulness of having a compatible scoping structure with snapshots...

Reply:
That's a compelling enough reason to have `schema`, `protocol` , and `recordIds` be mutually exclusive. How about this: `schema`, `protocol` , and `recordIds` are mutually exclusive. When `protocols` is present, one or zero of `contextId` or `protocolPath` may be present.

Reply:
Update from live discussion earlier today:

With a change to Snapshots, it's no longer necessary that grants and snapshots have the same scope structure. So, there's no need to make `protocols`, `schema`, and `recordIds` mutually exclusive. It's still probably best practice for app developers to use only one at a time.

The change to Snapshots is that we have separate snapshot scopes for protocols and grants, and those scopes may overlap, meaning that a single record may fall into both a protocol and grant snapshot scope.

Reply:
Following offline discussion, the above is probably mostly obsolete now, with `recordId` possibly being yet another intersecting branch.